# Import libraries

In [1]:
from sklearn.neighbors import KNeighborsClassifier
#from PIL import Image
import numpy as np
from sklearn.metrics import accuracy_score
import tensorflow as tf
#from keras.layers import *
from tensorflow.keras import *
from tensorflow.keras.layers import *
from tensorflow.keras.preprocessing.image import array_to_img, img_to_array, load_img
import np_utils
from tensorflow.keras.utils import to_categorical
from tensorflow import lite


# Load dataset

In [2]:
word={'a':0,'b':1,'c':2,'d':3,'e':4,'f':5,'g':6,'h':7,'i':8,'j':9,'k':10,'l':11,'m':12,'n':13,'o':14,'p':15,'q':16,'r':17,'s':18,'t':19,'u':20,'v':21,'w':22,'x':23,'y':24,'z':25}

### User1 dataset

In [3]:
data_x_1=[]
data_y_1=[]
test_x_1=[]
test_y_1=[]

cnt1=0 #user1 test데이터 수

for j in range(0,100):
    for i in ['a','b','c','d','e','f','g','h','i','j','k','l','m','n','o','p','q','r','s','t','u','v','w','x','y','z']:

       
       
        
        try:
            f = open("matrixWav2\\cs\\"+i+"\\"+i+str(j)+".txt")
            line = f.readline()
            temp=line.split(",")
            f.close()
        except:
            continue
        
        
        temp_np=np.array(list(map(float, temp)))
        #temp_np=(temp_np - temp_np.min(axis=0)) / (temp_np.max(axis=0) - temp_np.min(axis=0)) #min-max scaling
        temp_np=temp_np.reshape(1248,24,1) #not resize
        
        
        if j in range(80,100):
            cnt1+=1
            test_x_1.append(temp_np)
            test_y_1.append(word[i])
        
        else:
            data_x_1.append(temp_np)
            data_y_1.append(word[i])            

            
test_x_1=np.array(test_x_1)
test_y_1=np.array(test_y_1)
data_x_1=np.array(data_x_1)
data_y_1=np.array(data_y_1)

#one hot encoding은 나중에.
#test_y_1 = to_categorical(test_y_1,num_classes=26)
#data_y_1 = to_categorical(data_y_1,num_classes=26)

In [4]:
# 사이킷런 kNN은 n차원 데이터 읽기 x -> 변환
data_x_1= data_x_1.reshape(data_x_1.shape[0],-1)
test_x_1= test_x_1.reshape(test_x_1.shape[0],-1)

In [5]:
print(data_x_1.shape)
print(data_y_1.shape)
print(test_x_1.shape)
print(test_y_1.shape)

(2064, 29952)
(2064,)
(520, 29952)
(520,)


### User2 dataset

In [34]:
data_x_2=[]
data_y_2=[]
test_x_2=[]
test_y_2=[]


cnt2=0

for j in range(0,100):
    for i in ['a','b','c','d','e','f','g','h','i','j','k','l','m','n','o','p','q','r','s','t','u','v','w','x','y','z']:
        #print(i+str(j)+".txt")
       
       
            
        
        
        try:
            f = open("matrixWav2\\yj\\"+i+"\\"+i+str(j)+".txt")
            line = f.readline()
            temp=line.split(",")
            f.close()
        except:
            continue
        
        
        temp_np=np.array(list(map(float, temp)))
        #temp_np=(temp_np - temp_np.min(axis=0)) / (temp_np.max(axis=0) - temp_np.min(axis=0)) #min-max scaling
        temp_np=temp_np.reshape(1248,24,1) #not resize
        
        
        if j in range(80,100):
            cnt2+=1
            test_x_2.append(temp_np)
            test_y_2.append(word[i])
        
        else:
            data_x_2.append(temp_np)
            data_y_2.append(word[i])            

            
test_x_2=np.array(test_x_2)
test_y_2=np.array(test_y_2)
data_x_2=np.array(data_x_2)
data_y_2=np.array(data_y_2)

# one hot encoding은 나중에
#test_y_2 = to_categorical(test_y_2,num_classes=26)
#data_y_2 = to_categorical(data_y_2,num_classes=26)

In [35]:
# 사이킷런 kNN은 n차원 데이터 읽기 x -> 변환
data_x_2= data_x_2.reshape(data_x_2.shape[0],-1)
test_x_2= test_x_2.reshape(test_x_2.shape[0],-1)

In [36]:
print(data_x_2.shape)
print(data_y_2.shape)
print(test_x_2.shape)
print(test_y_2.shape)

(2011, 29952)
(2011,)
(495, 29952)
(495,)


### Merge data

In [37]:
data_x=np.concatenate((data_x_1,data_x_2),axis=0)
data_y=np.concatenate((data_y_1,data_y_2),axis=0)
test_x=np.concatenate((test_x_1,test_x_2),axis=0)
test_y=np.concatenate((test_y_1,test_y_2),axis=0)

In [38]:
print(data_x.shape)
print(data_y.shape)
print(test_x.shape)
print(test_y.shape)

(4075, 29952)
(4075,)
(1015, 29952)
(1015,)


# Scikit learn 

## Train

### User1

In [14]:
neigh1 = KNeighborsClassifier(n_neighbors=4)
neigh1.fit(data_x_1,data_y_1)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=4, p=2,
                     weights='uniform')

### User2

In [17]:
neigh2 = KNeighborsClassifier(n_neighbors=4)
neigh2.fit(data_x_2,data_y_2)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=4, p=2,
                     weights='uniform')

### All

In [20]:
neigh = KNeighborsClassifier(n_neighbors=4)
neigh.fit(data_x,data_y)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=4, p=2,
                     weights='uniform')

## Test

### User1

In [15]:
pred_y1=neigh1.predict(test_x_1)

### User2

In [18]:
pred_y2=neigh2.predict(test_x_2)

### All

In [22]:
pred_y=neigh.predict(test_x)

## Compute accuracy

### User1

In [16]:
accuracy_score(test_y_1,pred_y1)

0.9288461538461539

### Label accuracy in User1

In [52]:
alpha_result=list([0]*26)
alpha_cnt=list([0]*26)
for idx,i in enumerate(test_y_1):
    alpha_cnt[i]+=1
    if i==pred_y1[idx]:
        alpha_result[i]+=1
alpha_acc={chr(i):0 for i in range(ord('a'),ord('z')+1)}
for idx,i in enumerate(alpha_result):
    if(alpha_cnt[idx]!=0):
        alpha_acc[chr(idx+ord('a'))]=(alpha_result[idx]/alpha_cnt[idx])
print(alpha_acc)

{'a': 0.85, 'b': 0.6, 'c': 1.0, 'd': 1.0, 'e': 0.8, 'f': 0.9, 'g': 0.95, 'h': 0.9, 'i': 0.9, 'j': 0.9, 'k': 1.0, 'l': 0.8, 'm': 1.0, 'n': 1.0, 'o': 1.0, 'p': 1.0, 'q': 1.0, 'r': 1.0, 's': 1.0, 't': 0.8, 'u': 0.8, 'v': 1.0, 'w': 1.0, 'x': 1.0, 'y': 1.0, 'z': 0.95}


### User2

In [19]:
accuracy_score(test_y_2,pred_y2)

0.8303030303030303

### Label accuracy in User2

In [53]:
alpha_result=list([0]*26)
alpha_cnt=list([0]*26)
for idx,i in enumerate(test_y_2):
    alpha_cnt[i]+=1
    if i==pred_y2[idx]:
        alpha_result[i]+=1
alpha_acc={chr(i):0 for i in range(ord('a'),ord('z')+1)}
for idx,i in enumerate(alpha_result):
    if(alpha_cnt[idx]!=0):
        alpha_acc[chr(idx+ord('a'))]=(alpha_result[idx]/alpha_cnt[idx])
print(alpha_acc)

{'a': 1.0, 'b': 1.0, 'c': 0.9, 'd': 1.0, 'e': 0.65, 'f': 0.95, 'g': 0.7368421052631579, 'h': 0.85, 'i': 0.6111111111111112, 'j': 0.7, 'k': 1.0, 'l': 0.95, 'm': 0.95, 'n': 0.85, 'o': 0.8, 'p': 0.75, 'q': 0.8947368421052632, 'r': 0.9, 's': 0.6363636363636364, 't': 0.9, 'u': 0.65, 'v': 0.25, 'w': 0.7, 'x': 0.95, 'y': 1.0, 'z': 0.85}


### All

In [23]:
accuracy_score(test_y,pred_y)

0.8817733990147784

### User accuracy in All data

In [24]:
result1=0
result2=0
#result3=0

for i in enumerate(test_y):
    if i[1]!=pred_y[i[0]]:
        if(i[0] in range(0,cnt1)):
            result1+=1
        elif(i[0] in range(cnt1,cnt1+cnt2)):
            result2+=1
        #elif(i[0] in range (cnt1+cnt2,cnt1+cnt2+cnt3)):
            #result3+=1
            
result1=((cnt1-result1)/cnt1)*100
result2=((cnt2-result2)/cnt2)*100
#result3=((cnt3-result3)/cnt3)*100

print(result1,result2) #,result3)

92.5 83.63636363636363


### Label accuracy in All data

In [25]:
alpha_result=list([0]*26)
alpha_cnt=list([0]*26)
for idx,i in enumerate(test_y):
    alpha_cnt[i]+=1
    if i==pred_y[idx]:
        alpha_result[i]+=1
alpha_acc={chr(i):0 for i in range(ord('a'),ord('z')+1)}
for idx,i in enumerate(alpha_result):
    if(alpha_cnt[idx]!=0):
        alpha_acc[chr(idx+ord('a'))]=(alpha_result[idx]/alpha_cnt[idx])
print(alpha_acc)

{'a': 0.8157894736842105, 'b': 0.9, 'c': 0.9, 'd': 1.0, 'e': 0.725, 'f': 0.9, 'g': 0.8717948717948718, 'h': 0.875, 'i': 0.7631578947368421, 'j': 0.8666666666666667, 'k': 1.0, 'l': 0.85, 'm': 0.975, 'n': 0.95, 'o': 0.9, 'p': 0.9, 'q': 0.8974358974358975, 'r': 0.95, 's': 0.8709677419354839, 't': 0.85, 'u': 0.7, 'v': 0.6, 'w': 1.0, 'x': 0.975, 'y': 0.975, 'z': 0.9}


### Result in All data

In [26]:
for idx,i in enumerate(test_y):
    if i!=pred_y[idx]:
        print("[",idx,'] real:',list(word.keys())[i],'| predict:',list(word.keys())[pred_y[i]])

[ 0 ] real: a | predict: d
[ 5 ] real: f | predict: h
[ 7 ] real: h | predict: l
[ 30 ] real: e | predict: e
[ 52 ] real: a | predict: d
[ 56 ] real: e | predict: e
[ 78 ] real: a | predict: d
[ 83 ] real: f | predict: h
[ 99 ] real: v | predict: v
[ 130 ] real: a | predict: d
[ 172 ] real: q | predict: q
[ 182 ] real: a | predict: d
[ 183 ] real: b | predict: b
[ 189 ] real: h | predict: l
[ 190 ] real: i | predict: i
[ 234 ] real: a | predict: d
[ 235 ] real: b | predict: b
[ 240 ] real: g | predict: g
[ 254 ] real: u | predict: u
[ 269 ] real: j | predict: j
[ 279 ] real: t | predict: t
[ 284 ] real: y | predict: y
[ 285 ] real: z | predict: z
[ 294 ] real: i | predict: i
[ 295 ] real: j | predict: j
[ 306 ] real: u | predict: u
[ 339 ] real: b | predict: b
[ 368 ] real: e | predict: e
[ 383 ] real: t | predict: t
[ 384 ] real: u | predict: u
[ 390 ] real: a | predict: d
[ 409 ] real: t | predict: t
[ 420 ] real: e | predict: e
[ 453 ] real: l | predict: l
[ 469 ] real: b | predict:

# Tensorflow

Reference : https://hubring.tistory.com/23

### Select data

In [41]:
data_train_x=data_x_1
data_train_y=to_categorical(data_y_1,num_classes=26)
data_test_x=test_x_1
data_test_y=to_categorical(test_y_1,num_classes=26)

### Setting

In [27]:
import tensorflow.compat.v1 as tf 
tf.disable_v2_behavior()

Instructions for updating:
non-resource variables are not supported in the long term


### Train & Predict

In [49]:
# region k-NN algorithm 

# x_data_train : training data 1248*24=29952
# y_data_train : training data one hot encoding label
# x_data_test : test data 1248*24=29952
x_data_train = tf.placeholder(tf.float64, [None, 29952])
y_data_train = tf.placeholder(tf.int64,[None, 26])
x_data_test = tf.placeholder(tf.float64, [29952])
paramK = tf.placeholder(tf.int32)

# distance : K-Nearest Neighbor - 오차 거리를 구한다(유클리드 거리)
distance = tf.reduce_sum(tf.abs(tf.add(x_data_train, tf.negative(x_data_test))), reduction_indices=1)

# nearest k points
_, top_k_indices = tf.nn.top_k(tf.negative(distance), k=paramK) #가장 가까운 K개의 인덱스를 가져온다. 
top_k_label = tf.gather(y_data_train, top_k_indices)  #Index에 해당하는 숫자 라벨를 가져온다.

sum_up_predictions = tf.reduce_sum(top_k_label, axis=0) #공통되는 숫자 라벨의 개수를 더한다. 
prediction = tf.argmax(sum_up_predictions) #가장 가까운 숫자값을 구한다.

# endregion


In [50]:
# Tensorflow 세션 실행
init = tf.initialize_all_variables()

sess = tf.Session()
sess.run(init)


# 정확도를 초기화
accuracy = 0
confusion_matrix = [[0 for i in range(26)] for j in range(26)]


for i in range(len(data_test_x)):

    # k-NN 알고리즘 수행 예측된 숫자를 구한다.
    pre_label = sess.run(prediction, feed_dict={x_data_train: data_train_x, y_data_train: data_train_y, x_data_test: data_test_x[i,:], paramK :  4})
    real_label = np.argmax(data_test_y[i])

    #nn_index의 라벨값과 실제 라벨값을 비교한다.
    print("Iteration : ", i,"Real label : ", real_label,"Predict label : ", pre_label)

    confusion_matrix[real_label][pre_label] +=1

    # 예측도 파악
    # KNN은 비교 데이터에서 가장 가까운 것을 찾는 것이므로 매번 확률을 갱신해야한다(가중치를 찾는게 아니다)
    # 가장 가까운 것이 무엇이 될지 모름
    if pre_label == real_label :
        accuracy += 1./len(data_test_x)
   





Iteration :  0 Real label :  0 Predict label :  20
Iteration :  1 Real label :  1 Predict label :  1
Iteration :  2 Real label :  2 Predict label :  2
Iteration :  3 Real label :  3 Predict label :  3
Iteration :  4 Real label :  4 Predict label :  4
Iteration :  5 Real label :  5 Predict label :  8
Iteration :  6 Real label :  6 Predict label :  6
Iteration :  7 Real label :  7 Predict label :  7
Iteration :  8 Real label :  8 Predict label :  8
Iteration :  9 Real label :  9 Predict label :  9
Iteration :  10 Real label :  10 Predict label :  10
Iteration :  11 Real label :  11 Predict label :  11
Iteration :  12 Real label :  12 Predict label :  12
Iteration :  13 Real label :  13 Predict label :  13
Iteration :  14 Real label :  14 Predict label :  14
Iteration :  15 Real label :  15 Predict label :  15
Iteration :  16 Real label :  16 Predict label :  16
Iteration :  17 Real label :  17 Predict label :  17
Iteration :  18 Real label :  18 Predict label :  18
Iteration :  19 Real l

Iteration :  156 Real label :  0 Predict label :  0
Iteration :  157 Real label :  1 Predict label :  1
Iteration :  158 Real label :  2 Predict label :  2
Iteration :  159 Real label :  3 Predict label :  3
Iteration :  160 Real label :  4 Predict label :  4
Iteration :  161 Real label :  5 Predict label :  5
Iteration :  162 Real label :  6 Predict label :  6
Iteration :  163 Real label :  7 Predict label :  7
Iteration :  164 Real label :  8 Predict label :  8
Iteration :  165 Real label :  9 Predict label :  9
Iteration :  166 Real label :  10 Predict label :  10
Iteration :  167 Real label :  11 Predict label :  11
Iteration :  168 Real label :  12 Predict label :  12
Iteration :  169 Real label :  13 Predict label :  13
Iteration :  170 Real label :  14 Predict label :  14
Iteration :  171 Real label :  15 Predict label :  15
Iteration :  172 Real label :  16 Predict label :  16
Iteration :  173 Real label :  17 Predict label :  17
Iteration :  174 Real label :  18 Predict label 

Iteration :  310 Real label :  24 Predict label :  24
Iteration :  311 Real label :  25 Predict label :  25
Iteration :  312 Real label :  0 Predict label :  0
Iteration :  313 Real label :  1 Predict label :  3
Iteration :  314 Real label :  2 Predict label :  2
Iteration :  315 Real label :  3 Predict label :  3
Iteration :  316 Real label :  4 Predict label :  4
Iteration :  317 Real label :  5 Predict label :  5
Iteration :  318 Real label :  6 Predict label :  6
Iteration :  319 Real label :  7 Predict label :  7
Iteration :  320 Real label :  8 Predict label :  8
Iteration :  321 Real label :  9 Predict label :  9
Iteration :  322 Real label :  10 Predict label :  10
Iteration :  323 Real label :  11 Predict label :  11
Iteration :  324 Real label :  12 Predict label :  12
Iteration :  325 Real label :  13 Predict label :  13
Iteration :  326 Real label :  14 Predict label :  14
Iteration :  327 Real label :  15 Predict label :  15
Iteration :  328 Real label :  16 Predict label 

Iteration :  464 Real label :  22 Predict label :  22
Iteration :  465 Real label :  23 Predict label :  23
Iteration :  466 Real label :  24 Predict label :  24
Iteration :  467 Real label :  25 Predict label :  25
Iteration :  468 Real label :  0 Predict label :  0
Iteration :  469 Real label :  1 Predict label :  9
Iteration :  470 Real label :  2 Predict label :  2
Iteration :  471 Real label :  3 Predict label :  3
Iteration :  472 Real label :  4 Predict label :  4
Iteration :  473 Real label :  5 Predict label :  5
Iteration :  474 Real label :  6 Predict label :  6
Iteration :  475 Real label :  7 Predict label :  7
Iteration :  476 Real label :  8 Predict label :  8
Iteration :  477 Real label :  9 Predict label :  9
Iteration :  478 Real label :  10 Predict label :  10
Iteration :  479 Real label :  11 Predict label :  21
Iteration :  480 Real label :  12 Predict label :  12
Iteration :  481 Real label :  13 Predict label :  13
Iteration :  482 Real label :  14 Predict label 

In [51]:

print("Accuracy : ", round(accuracy*100, 2)," %")
print("The resulting confusion matrix")
# confusion matrix 생성
for i in range(26):
    print(chr(ord('a')+i),confusion_matrix[i],"acc = ",confusion_matrix[i][i]/np.sum(confusion_matrix[i]))

Accuracy :  92.69  %
The resulting confusion matrix
a [16, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 0, 0] acc =  0.8
b [1, 13, 0, 4, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] acc =  0.65
c [0, 0, 20, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] acc =  1.0
d [0, 0, 0, 20, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] acc =  1.0
e [0, 0, 4, 0, 16, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] acc =  0.8
f [0, 0, 0, 0, 0, 18, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] acc =  0.9
g [0, 0, 0, 0, 0, 0, 19, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] acc =  0.95
h [0, 0, 0, 0, 0, 0, 0, 19, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] acc =  0.95
i [0, 0, 0, 0, 0, 0, 0, 0, 17, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] acc =  0.85
j [0, 0, 0, 0, 0, 0, 0, 1, 0, 18, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] acc =  0.9
k [0, 0, 0, 0,

In [63]:
pre_label

25

## Convert session to tflite 

In [ ]:
converter = tf.lite.TFLiteConverter.from_session(sess, input_tensors=model.inputs, output_tensors=model.outputs)

In [ ]:
converter = tf.compat.v1.lite.TFLiteConverter.from_session(sess, input_tensors=model.inputs, output_tensors=model.outputs)

In [61]:
input_ts=list({x_data_train: data_train_x, y_data_train: data_train_y, x_data_test: data_test_x[i,:], paramK :  4}.values())

In [69]:
input_ts2=list([x_data_train,y_data_train,x_data_test,paramK])

In [73]:
len(input_ts2)

4

In [71]:
input_ts2[0]

<tf.Tensor 'Placeholder_8:0' shape=(?, 29952) dtype=float64>

In [72]:
input_ts2[1]

<tf.Tensor 'Placeholder_9:0' shape=(?, 26) dtype=int64>

# 개선점

input tensor 모양을 [<tf.Tensor 'input_1:0' shape=(?, 29952) dtype=float32>] 이런식으로 만들어줘야함..

-> tensorflow code 수정?

# Version2

기존 구조는 매번 천단위 train 데이터와 1개의 test데이터 사이의 거리를 구해서 상위4개의 라벨을 바탕으로 예측하는 것

각 label별로 평균 중심점을 구해서 26개의 데이터로 축소 -> 테스트 데이터가 들어오면 26개의 데이터와의 거리를 바탕으로 예측

기존 코드보다 정확도가 얼마나 떨어질지가 의문..

## Data separation

### User1

In [6]:
labels1=[]
for i in range(26):
    temp=[]
    for idx,j in enumerate(data_y_1):
        if(j==i):
            temp.append(data_x_1[idx].tolist())
    labels1.append(temp[:])

In [7]:
len(labels1)

26

In [9]:
for i in range(26):
    print(i,":",len(labels1[i]))

0 : 79
1 : 80
2 : 80
3 : 80
4 : 80
5 : 80
6 : 69
7 : 80
8 : 80
9 : 78
10 : 79
11 : 80
12 : 79
13 : 80
14 : 80
15 : 80
16 : 80
17 : 80
18 : 80
19 : 80
20 : 80
21 : 80
22 : 80
23 : 80
24 : 80
25 : 80


In [17]:
centroid1=[]
for i in range(26):
    centroid1.append(np.mean(labels1[i],axis=0))

In [21]:
centroid1

[array([-42.86217584, -42.47692695, -42.61891138, ..., -70.87655125,
        -70.95061299, -71.04935618]),
 array([-45.61665409, -45.61422906, -45.46531947, ..., -71.43911869,
        -71.40485322, -71.46578341]),
 array([-42.90768357, -43.75992257, -43.6384705 , ..., -70.59286605,
        -70.73431345, -70.59190534]),
 array([-42.48514159, -42.59611954, -41.52035784, ..., -72.83279681,
        -72.10221544, -71.73175501]),
 array([-42.35030881, -42.15176853, -43.3280671 , ..., -64.92977454,
        -65.02153623, -65.19979825]),
 array([-42.95519379, -42.79909938, -42.72239799, ..., -70.85547288,
        -71.08081786, -71.18113984]),
 array([-42.58818238, -41.63993499, -41.84129794, ..., -69.57501906,
        -69.55581959, -69.56153999]),
 array([-43.682846  , -42.01528447, -41.80389671, ..., -69.38629054,
        -69.64974727, -70.10450204]),
 array([-47.19297424, -47.31687435, -46.85571981, ..., -66.78839686,
        -66.98810664, -67.20300125]),
 array([-47.18163347, -46.86457462, -

In [22]:
centroid_y1=[i for i in range(26)]

### User2

In [39]:
labels2=[]
for i in range(26):
    temp=[]
    for idx,j in enumerate(data_y_2):
        if(j==i):
            temp.append(data_x_2[idx].tolist())
    labels2.append(temp[:])

In [40]:
len(labels2)

26

In [41]:
for i in range(26):
    print(i,":",len(labels2[i]))

0 : 80
1 : 80
2 : 80
3 : 80
4 : 79
5 : 78
6 : 76
7 : 79
8 : 72
9 : 60
10 : 80
11 : 80
12 : 80
13 : 80
14 : 80
15 : 79
16 : 80
17 : 79
18 : 57
19 : 79
20 : 79
21 : 80
22 : 80
23 : 79
24 : 75
25 : 80


In [42]:
centroid2=[]
for i in range(26):
    centroid2.append(np.mean(labels2[i],axis=0))

In [43]:
centroid2

[array([-43.8920859 , -44.38233355, -43.78554054, ..., -66.9154869 ,
        -67.34931844, -68.03141804]),
 array([-45.20289851, -45.11429259, -44.95387199, ..., -71.67065816,
        -71.74056665, -71.54488387]),
 array([-44.24947489, -45.46022049, -43.4073073 , ..., -67.98905174,
        -68.06861525, -68.16018666]),
 array([-44.4866901 , -43.12031532, -41.53627372, ..., -68.09965367,
        -68.2997576 , -68.46058899]),
 array([-44.90071976, -44.89127257, -43.21013443, ..., -68.04947366,
        -68.0525802 , -68.14051711]),
 array([-45.03155655, -44.64295776, -46.31460282, ..., -65.2887285 ,
        -65.17205415, -65.24858058]),
 array([-43.05971662, -43.16851932, -43.78797351, ..., -67.61119933,
        -67.734529  , -67.83946762]),
 array([-47.68778728, -47.19526923, -46.45311856, ..., -69.95841838,
        -69.83248947, -69.79957466]),
 array([-46.52651794, -46.56816367, -47.47974038, ..., -69.12664144,
        -69.05840939, -69.04967775]),
 array([-47.1620001 , -46.96193813, -

In [44]:
centroid_y2=[i for i in range(26)]

### All

In [48]:
labels=[]
for i in range(26):
    temp=[]
    for idx,j in enumerate(data_y):
        if(j==i):
            temp.append(data_x[idx].tolist())
    labels.append(temp[:])

In [49]:
len(labels)

26

In [50]:
for i in range(26):
    print(i,":",len(labels[i]))

0 : 159
1 : 160
2 : 160
3 : 160
4 : 159
5 : 158
6 : 145
7 : 159
8 : 152
9 : 138
10 : 159
11 : 160
12 : 159
13 : 160
14 : 160
15 : 159
16 : 160
17 : 159
18 : 137
19 : 159
20 : 159
21 : 160
22 : 160
23 : 159
24 : 155
25 : 160


In [51]:
centroid=[]
for i in range(26):
    centroid.append(np.mean(labels[i],axis=0))

In [52]:
centroid

[array([-43.38036958, -43.43562209, -43.2058946 , ..., -68.8835629 ,
        -69.13864089, -69.53089674]),
 array([-45.4097763 , -45.36426082, -45.20959573, ..., -71.55488842,
        -71.57270994, -71.50533364]),
 array([-43.57857923, -44.61007153, -43.5228889 , ..., -69.29095889,
        -69.40146435, -69.376046  ]),
 array([-43.48591584, -42.85821743, -41.52831578, ..., -70.46622524,
        -70.20098652, -70.096172  ]),
 array([-43.61749413, -43.51290575, -43.26947162, ..., -66.47981372,
        -66.52752663, -66.66091014]),
 array([-43.98023363, -43.70935858, -44.49576493, ..., -68.10733325,
        -68.16383325, -68.25240805]),
 array([-42.83533136, -42.44112401, -42.86162445, ..., -68.54570665,
        -68.60121211, -68.65893654]),
 array([-45.67272248, -44.58898759, -44.11388744, ..., -69.67055531,
        -69.74054371, -69.95299724]),
 array([-46.87728441, -46.96222192, -47.1513085 , ..., -67.8959864 ,
        -67.96877636, -68.07774275]),
 array([-47.17309722, -46.90690658, -

In [53]:
centroid_y=[i for i in range(26)]

## Select data

### User1

In [26]:
data_train_x=centroid1
data_train_y=to_categorical(centroid_y1,num_classes=26)
data_test_x=test_x_1
data_test_y=to_categorical(test_y_1,num_classes=26)

### User2

In [45]:
data_train_x=centroid2
data_train_y=to_categorical(centroid_y2,num_classes=26)
data_test_x=test_x_2
data_test_y=to_categorical(test_y_2,num_classes=26)

### All

In [54]:
data_train_x=centroid
data_train_y=to_categorical(centroid_y,num_classes=26)
data_test_x=test_x
data_test_y=to_categorical(test_y,num_classes=26)

### User1's small data

In [63]:
data_train_x=data_x_1[:260]
data_train_y=to_categorical(data_y_1[:260],num_classes=26)
data_test_x=test_x_1
data_test_y=to_categorical(test_y_1,num_classes=26)

## Setting

In [27]:
import tensorflow.compat.v1 as tf 
tf.disable_v2_behavior()

Instructions for updating:
non-resource variables are not supported in the long term


## Train & Predict

In [28]:
# region k-NN algorithm 

# x_data_train : training data 1248*24=29952
# y_data_train : training data one hot encoding label
# x_data_test : test data 1248*24=29952
x_data_train = tf.placeholder(tf.float64, [None, 29952])
y_data_train = tf.placeholder(tf.int64,[None, 26])
x_data_test = tf.placeholder(tf.float64, [29952])
paramK = tf.placeholder(tf.int32)

# distance : K-Nearest Neighbor - 오차 거리를 구한다(유클리드 거리)
distance = tf.reduce_sum(tf.abs(tf.add(x_data_train, tf.negative(x_data_test))), reduction_indices=1)

# nearest k points
_, top_k_indices = tf.nn.top_k(tf.negative(distance), k=paramK) #가장 가까운 K개의 인덱스를 가져온다. 
top_k_label = tf.gather(y_data_train, top_k_indices)  #Index에 해당하는 숫자 라벨를 가져온다.

sum_up_predictions = tf.reduce_sum(top_k_label, axis=0) #공통되는 숫자 라벨의 개수를 더한다. 
prediction = tf.argmax(sum_up_predictions) #가장 가까운 숫자값을 구한다.

# endregion


In [68]:
# Tensorflow 세션 실행
init = tf.initialize_all_variables()

sess = tf.Session()
sess.run(init)


# 정확도를 초기화
accuracy = 0
confusion_matrix = [[0 for i in range(26)] for j in range(26)]


for i in range(len(data_test_x)):

    # k-NN 알고리즘 수행 예측된 숫자를 구한다.
    pre_label = sess.run(prediction, feed_dict={x_data_train: data_train_x, y_data_train: data_train_y, x_data_test: data_test_x[i,:], paramK :  1})
    real_label = np.argmax(data_test_y[i])

    #nn_index의 라벨값과 실제 라벨값을 비교한다.
    print("Iteration : ", i,"Real label : ", real_label,"Predict label : ", pre_label)

    confusion_matrix[real_label][pre_label] +=1

    # 예측도 파악
    # KNN은 비교 데이터에서 가장 가까운 것을 찾는 것이므로 매번 확률을 갱신해야한다(가중치를 찾는게 아니다)
    # 가장 가까운 것이 무엇이 될지 모름
    if pre_label == real_label :
        accuracy += 1./len(data_test_x)
   





Iteration :  0 Real label :  0 Predict label :  14
Iteration :  1 Real label :  1 Predict label :  3
Iteration :  2 Real label :  2 Predict label :  2
Iteration :  3 Real label :  3 Predict label :  3
Iteration :  4 Real label :  4 Predict label :  4
Iteration :  5 Real label :  5 Predict label :  8
Iteration :  6 Real label :  6 Predict label :  6
Iteration :  7 Real label :  7 Predict label :  11
Iteration :  8 Real label :  8 Predict label :  11
Iteration :  9 Real label :  9 Predict label :  7
Iteration :  10 Real label :  10 Predict label :  15
Iteration :  11 Real label :  11 Predict label :  21
Iteration :  12 Real label :  12 Predict label :  16
Iteration :  13 Real label :  13 Predict label :  13
Iteration :  14 Real label :  14 Predict label :  14
Iteration :  15 Real label :  15 Predict label :  15
Iteration :  16 Real label :  16 Predict label :  16
Iteration :  17 Real label :  17 Predict label :  17
Iteration :  18 Real label :  18 Predict label :  17
Iteration :  19 Real

Iteration :  159 Real label :  3 Predict label :  24
Iteration :  160 Real label :  4 Predict label :  0
Iteration :  161 Real label :  5 Predict label :  8
Iteration :  162 Real label :  6 Predict label :  6
Iteration :  163 Real label :  7 Predict label :  1
Iteration :  164 Real label :  8 Predict label :  11
Iteration :  165 Real label :  9 Predict label :  7
Iteration :  166 Real label :  10 Predict label :  15
Iteration :  167 Real label :  11 Predict label :  21
Iteration :  168 Real label :  12 Predict label :  16
Iteration :  169 Real label :  13 Predict label :  13
Iteration :  170 Real label :  14 Predict label :  14
Iteration :  171 Real label :  15 Predict label :  15
Iteration :  172 Real label :  16 Predict label :  16
Iteration :  173 Real label :  17 Predict label :  17
Iteration :  174 Real label :  18 Predict label :  18
Iteration :  175 Real label :  19 Predict label :  19
Iteration :  176 Real label :  20 Predict label :  14
Iteration :  177 Real label :  21 Predic

Iteration :  316 Real label :  4 Predict label :  4
Iteration :  317 Real label :  5 Predict label :  0
Iteration :  318 Real label :  6 Predict label :  5
Iteration :  319 Real label :  7 Predict label :  1
Iteration :  320 Real label :  8 Predict label :  11
Iteration :  321 Real label :  9 Predict label :  0
Iteration :  322 Real label :  10 Predict label :  10
Iteration :  323 Real label :  11 Predict label :  21
Iteration :  324 Real label :  12 Predict label :  16
Iteration :  325 Real label :  13 Predict label :  13
Iteration :  326 Real label :  14 Predict label :  14
Iteration :  327 Real label :  15 Predict label :  15
Iteration :  328 Real label :  16 Predict label :  16
Iteration :  329 Real label :  17 Predict label :  17
Iteration :  330 Real label :  18 Predict label :  18
Iteration :  331 Real label :  19 Predict label :  19
Iteration :  332 Real label :  20 Predict label :  18
Iteration :  333 Real label :  21 Predict label :  21
Iteration :  334 Real label :  22 Predi

Iteration :  472 Real label :  4 Predict label :  3
Iteration :  473 Real label :  5 Predict label :  5
Iteration :  474 Real label :  6 Predict label :  16
Iteration :  475 Real label :  7 Predict label :  1
Iteration :  476 Real label :  8 Predict label :  11
Iteration :  477 Real label :  9 Predict label :  1
Iteration :  478 Real label :  10 Predict label :  15
Iteration :  479 Real label :  11 Predict label :  21
Iteration :  480 Real label :  12 Predict label :  16
Iteration :  481 Real label :  13 Predict label :  17
Iteration :  482 Real label :  14 Predict label :  14
Iteration :  483 Real label :  15 Predict label :  15
Iteration :  484 Real label :  16 Predict label :  16
Iteration :  485 Real label :  17 Predict label :  17
Iteration :  486 Real label :  18 Predict label :  18
Iteration :  487 Real label :  19 Predict label :  19
Iteration :  488 Real label :  20 Predict label :  17
Iteration :  489 Real label :  21 Predict label :  21
Iteration :  490 Real label :  22 Pred

## Result

### User1

In [32]:

print("Accuracy : ", round(accuracy*100, 2)," %")
print("The resulting confusion matrix")
# confusion matrix 생성
for i in range(26):
    print(chr(ord('a')+i),confusion_matrix[i],"acc = ",confusion_matrix[i][i]/np.sum(confusion_matrix[i]))

Accuracy :  78.08  %
The resulting confusion matrix
a [7, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 13, 0, 0, 0, 0, 0] acc =  0.35
b [0, 11, 0, 4, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0] acc =  0.55
c [0, 0, 20, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] acc =  1.0
d [0, 0, 0, 16, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0] acc =  0.8
e [3, 5, 9, 1, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] acc =  0.1
f [0, 0, 0, 0, 0, 12, 0, 0, 2, 6, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] acc =  0.6
g [0, 0, 0, 1, 0, 0, 6, 0, 1, 3, 5, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0] acc =  0.3
h [0, 0, 0, 1, 0, 0, 1, 15, 0, 1, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] acc =  0.75
i [0, 0, 0, 0, 0, 0, 0, 0, 18, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] acc =  0.9
j [0, 0, 1, 0, 0, 0, 0, 0, 0, 19, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] acc =  0.95
k [0, 0, 0, 0, 0

### User2

In [47]:

print("Accuracy : ", round(accuracy*100, 2)," %")
print("The resulting confusion matrix")
# confusion matrix 생성
for i in range(26):
    print(chr(ord('a')+i),confusion_matrix[i],"acc = ",confusion_matrix[i][i]/np.sum(confusion_matrix[i]))

Accuracy :  57.98  %
The resulting confusion matrix
a [18, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] acc =  1.0
b [10, 10, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] acc =  0.5
c [0, 0, 5, 0, 1, 0, 0, 0, 0, 0, 0, 2, 0, 0, 12, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] acc =  0.25
d [0, 0, 0, 18, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0] acc =  0.9
e [0, 0, 10, 3, 0, 0, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] acc =  0.0
f [0, 0, 1, 0, 0, 11, 0, 1, 0, 0, 0, 0, 0, 2, 1, 0, 0, 1, 0, 3, 0, 0, 0, 0, 0, 0] acc =  0.55
g [0, 3, 0, 0, 0, 2, 5, 0, 0, 1, 0, 0, 0, 4, 0, 0, 2, 0, 0, 0, 0, 0, 1, 0, 1, 0] acc =  0.2631578947368421
h [0, 6, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 4, 0, 0, 0, 0, 0] acc =  0.35
i [0, 0, 0, 0, 0, 0, 0, 0, 11, 7, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] acc =  0.6111111111111112
j [0, 0, 1, 0, 0, 0, 1, 0, 1, 6, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0

### All

In [56]:

print("Accuracy : ", round(accuracy*100, 2)," %")
print("The resulting confusion matrix")
# confusion matrix 생성
for i in range(26):
    print(chr(ord('a')+i),confusion_matrix[i],"acc = ",confusion_matrix[i][i]/np.sum(confusion_matrix[i]))

Accuracy :  57.83  %
The resulting confusion matrix
a [17, 0, 0, 6, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 15, 0, 0, 0, 0, 0] acc =  0.4473684210526316
b [0, 20, 0, 14, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 3, 0, 0, 0] acc =  0.5
c [0, 0, 28, 0, 4, 0, 0, 0, 0, 0, 0, 1, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] acc =  0.7
d [0, 0, 0, 8, 0, 0, 5, 0, 0, 0, 0, 0, 0, 3, 4, 0, 5, 0, 0, 6, 6, 0, 3, 0, 0, 0] acc =  0.2
e [2, 1, 19, 7, 1, 0, 0, 0, 0, 0, 0, 7, 0, 0, 2, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0] acc =  0.025
f [0, 0, 0, 1, 1, 20, 0, 0, 2, 7, 0, 0, 0, 3, 0, 0, 0, 0, 1, 0, 1, 0, 0, 4, 0, 0] acc =  0.5
g [0, 1, 1, 2, 0, 3, 6, 0, 2, 8, 0, 0, 7, 5, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 1, 0] acc =  0.15384615384615385
h [1, 6, 0, 7, 0, 0, 2, 13, 0, 1, 0, 0, 0, 0, 1, 0, 0, 3, 0, 0, 5, 0, 1, 0, 0, 0] acc =  0.325
i [0, 0, 0, 0, 0, 0, 0, 0, 25, 11, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0] acc =  0.6578947368421053
j [0, 2, 1, 0, 0, 0, 1, 0, 6, 19, 0, 0, 0, 0, 0, 1, 0, 0, 0,

# 정확도 변화

|version|User1|User2|All|
|------|---|---|---|
|version1|0.92|0.83|0.88|
|version2|0.78|0.57|0.57|